js/state/syncUtil.js

```diff
@@ -52,57 +52,95 @@ const pickFields = (object, fields) => {
 }
 
 /**
- * Apply a bookmark or historySite SyncRecord to the browser data store.
- * @param {Object} record
+ * Apply bookmark or historySite SyncRecords to the browser data store.
+ * @param {Array.Object} records
  * @param {Immutable.Map} siteDetail
  */
-const applySiteRecord = (record) => {
+const applySiteRecords = (records) => {
   const appActions = require('../actions/appActions')
   const siteTags = require('../constants/siteTags')
-  const objectId = new Immutable.List(record.objectId)
-  const category = CATEGORY_MAP[record.objectData].categoryName
-  const existingObject = module.exports.getObjectById(objectId, category)
-  const existingObjectData = existingObject && existingObject[1]
-  let tag
-
-  let siteProps = Object.assign(
-    {},
-    existingObjectData && existingObjectData.toJS(),
-    record.historySite,
-    record.bookmark,
-    record.bookmark && record.bookmark.site,
-    {objectId}
-  )
-  if (record.objectData === 'bookmark') {
-    const existingFolderId = existingObjectData && existingObjectData.get('folderId')
-    if (existingFolderId) {
-      siteProps.folderId = existingFolderId
+
+  const batchedCreates = {} // map of tag to data
+  const updates = []
+  const deletes = []
+
+  records.forEach((record) => {
+    const objectId = new Immutable.List(record.objectId)
+    const category = CATEGORY_MAP[record.objectData].categoryName
+    const existingObject = module.exports.getObjectById(objectId, category)
+    const existingObjectData = existingObject && existingObject[1]
+    let tag
+
+    let siteProps = Object.assign(
+      {},
+      existingObjectData && existingObjectData.toJS(),
+      record.historySite,
+      record.bookmark,
+      record.bookmark && record.bookmark.site,
+      {objectId}
+    )
+    if (record.objectData === 'bookmark') {
+      const existingFolderId = existingObjectData && existingObjectData.get('folderId')
+      if (existingFolderId) {
+        siteProps.folderId = existingFolderId
+      }
+      const isFolder = (typeof siteProps.isFolder === 'boolean')
+        ? siteProps.isFolder
+        : !!existingFolderId
+      tag = isFolder
+        ? siteTags.BOOKMARK_FOLDER
+        : siteTags.BOOKMARK
+      const parentFolderObjectId = siteProps.parentFolderObjectId
+      if (parentFolderObjectId && parentFolderObjectId.length > 0) {
+        siteProps.parentFolderId =
+          getFolderIdByObjectId(new Immutable.List(parentFolderObjectId))
```

Comment:
for initial sync, if we apply all bookmarks and folders at once, this will always return null; so folders don't sync.

we could fix this by first applying folders and folderless bookmarks and folders (to ensure correct ordering); then applying foldered bookmarks. see below for more info